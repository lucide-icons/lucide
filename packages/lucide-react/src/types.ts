import { ReactSVG, SVGProps, ForwardRefExoticComponent, RefAttributes } from 'react';

export type IconNode = [elementName: keyof ReactSVG, attrs: Record<string, string>][];

export type SVGAttributes = Partial<SVGProps<SVGSVGElement>>;
type ElementAttributes = RefAttributes<SVGSVGElement> & SVGAttributes;

export interface LucideProps extends ElementAttributes {
  size?: string | number;
  absoluteStrokeWidth?: boolean;
}

export type LucideIcon = ForwardRefExoticComponent<
  Omit<LucideProps, 'ref'> & RefAttributes<SVGSVGElement>
>;
