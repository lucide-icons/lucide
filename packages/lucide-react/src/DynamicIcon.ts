'use client';

import { createElement, forwardRef, useEffect, useState } from 'react';
import { IconNode, LucideIcon, LucideProps } from './types';
import dynamicIconImports from './dynamicIconImports';
import Icon from './Icon';

export type DynamicIconModule = { default: LucideIcon; __iconNode: IconNode };

export type IconName = keyof typeof dynamicIconImports;

export const iconNames = Object.keys(dynamicIconImports) as Array<IconName>;

interface DynamicIconComponentProps extends LucideProps {
  name: IconName;
  fallback?: () => JSX.Element | null;
}

async function getIconNode(name: IconName) {
  if (!(name in dynamicIconImports)) {
    throw new Error('[lucide-react]: Name in Lucide DynamicIcon not found');
  }

  // TODO: Replace this with a generic iconNode package.
  const icon = (await dynamicIconImports[name]()) as DynamicIconModule;

  return icon.__iconNode;
}

/**
 * Dynamic Lucide icon component
 *
 * @component Icon
 * @param {object} props
 * @param {string} props.color - The color of the icon
 * @param {number} props.size - The size of the icon
 * @param {number} props.strokeWidth - The stroke width of the icon
 * @param {boolean} props.absoluteStrokeWidth - Whether to use absolute stroke width
 * @param {string} props.className - The class name of the icon
 * @param {IconNode} props.children - The children of the icon
 * @param {IconNode} props.iconNode - The icon node of the icon
 *
 * @returns {ForwardRefExoticComponent} LucideIcon
 */
const DynamicIcon = forwardRef<SVGSVGElement, DynamicIconComponentProps>(
  ({ name, fallback: Fallback, ...props }, ref) => {
    const [iconNode, setIconNode] = useState<IconNode>();

    useEffect(() => {
      getIconNode(name)
        .then(setIconNode)
        .catch((error) => {
          console.error(error);
        });
    }, [name]);

    if (iconNode == null) {
      if (Fallback == null) {
        return null;
      }

      return createElement(Fallback);
    }

    return createElement(Icon, {
      ref,
      ...props,
      iconNode,
    });
  },
);

export default DynamicIcon;
