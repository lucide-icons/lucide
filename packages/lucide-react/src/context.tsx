import { createContext, type ReactNode, useContext, useMemo } from 'react';
import { LucideProps } from './types';

type LucideConfig = {
  size: number;
  fill: string;
  color: string;
  strokeWidth: number;
  absoluteStrokeWidth: boolean;
};

const LucideContext = createContext<LucideProps>({});

type LucideProviderProps = {
  children: ReactNode;
} & Partial<LucideConfig>;

export function LucideProvider({
  children,
  size,
  fill,
  color,
  strokeWidth,
  absoluteStrokeWidth,
}: LucideProviderProps) {
  const value = useMemo(
    () => ({
      size,
      fill,
      color,
      strokeWidth,
      absoluteStrokeWidth,
    }),
    [size, fill, color, strokeWidth, absoluteStrokeWidth],
  );

  return <LucideContext.Provider value={value}>{children}</LucideContext.Provider>;
}

export const useLucideContext = () => useContext(LucideContext);
