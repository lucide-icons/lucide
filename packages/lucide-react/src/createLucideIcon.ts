import { forwardRef, createElement, ReactSVG, SVGProps } from 'react';
import defaultAttributes from './defaultAttributes';

type IconNode = [elementName: keyof ReactSVG, attrs: Record<string, string>][]

export type SVGAttributes = Partial<SVGProps<SVGSVGElement>>

export interface LucideProps extends SVGAttributes {
  size?: string | number
}
/**
 * Converts string to KebabCase
 * Copied from scripts/helper. If anyone knows how to properly import it here
 * then please fix it.
 *
 * @param {string} string
 * @returns {string} A kebabized string
 */
export const toKebabCase = (string: string) => string.replace(/([a-z0-9])([A-Z])/g, '$1-$2').toLowerCase();

const createLucideIcon = (iconName: string, iconNode: IconNode) => {
  const Component = forwardRef<SVGSVGElement, LucideProps>(
    ({ color = 'currentColor', size = 24, strokeWidth = 2, children, ...rest }, ref) =>
      createElement(
        'svg',
        {
          ref,
          ...defaultAttributes,
          width: size,
          height: size,
          stroke: color,
          strokeWidth,
          className: `lucide lucide-${toKebabCase(iconName)}`,
          ...rest,
        },
        [
          ...iconNode.map(([tag, attrs]) => createElement(tag, attrs)),
          ...(
            (Array.isArray(children) ? children : [children]) || []
          )
        ],
      ),
  );

  Component.displayName = `${iconName}`;

  return Component;
};

export default createLucideIcon
