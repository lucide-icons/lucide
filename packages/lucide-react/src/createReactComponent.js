import { forwardRef, createElement } from 'react';
import PropTypes from 'prop-types';

export default (iconName, [tag, attrs, children]) => {
  const Component = forwardRef(
    ({ color = 'currentColor', size = 24, strokeWidth = 2, ...rest }, ref) =>
      createElement(
        tag,
        {
          ref,
          ...attrs,
          width: size,
          height: size,
          stroke: color,
          strokeWidth,
          ...rest,
        },
        children.map(([childTag, childAttrs]) => createElement(childTag, childAttrs)),
      ),
  );

  Component.propTypes = {
    color: PropTypes.string,
    size: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    strokeWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  };

  Component.displayName = `${iconName}`;

  return Component;
};
