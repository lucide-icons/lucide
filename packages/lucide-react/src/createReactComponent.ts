import React, { forwardRef, createElement, ReactSVG, SVGProps } from 'react';
import PropTypes from 'prop-types';
import { IconNode, DEFAULT_ATTRIBUTES } from 'lucide'

export type SVGAttributes = Partial<SVGProps<SVGSVGElement>>

export interface LucideProps extends SVGAttributes {
  size?: string | number
}

/**
 * Converts string to KebabCase
 * Copied from scripts/helper. If anyone knows how to properly import it here
 * then please fix it.
 *
 * @param {string} string
 * @returns {string} A kebabized string
 */
export const toKebabCase = (string: string): string => string.replace(/([a-z0-9])([A-Z])/g, '$1-$2').toLowerCase();

/**
 * Converts string to CamelCase
 * Copied from scripts/helper. If anyone knows how to properly import it here
 * then please fix it.
 *
 * @param {string} string
 * @returns {string} A Camelized string
 */
export const toCamelCase = (string: string): string => string.replace(/-./g, x=>x[1].toUpperCase());

const defaultAttrs = Object.keys(DEFAULT_ATTRIBUTES).reduce((acc, key)=> {
  acc[toCamelCase(key)] = DEFAULT_ATTRIBUTES[key]
  return acc
}, {} as Record<string,string|number>)

export function createReactComponent (iconName: string, [_tag, DEFAULT_ATTRIBUTES, nodeChildren=[]]: IconNode) {
    const Component = forwardRef<SVGSVGElement, LucideProps>(
    ({ color = 'currentColor', size = 24, strokeWidth = 2, children, ...rest }, ref) =>
      createElement(
        'svg',
        {
          ref,
          ...defaultAttrs,
          width: size,
          height: size,
          stroke: color,
          strokeWidth,
          className: `lucide lucide-${toKebabCase(iconName)}`,
          ...rest,
        },
        [
          ...nodeChildren.map(([tag, attrs], index) => createElement(tag, { key: index, ...attrs })),
          ...(
            (Array.isArray(children) ? children : [children]) || []
          )
        ],
      ),
  );

  Component.propTypes = {
    color: PropTypes.string,
    size: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    strokeWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  };

  Component.displayName = `${iconName}`;

  return Component;
};
