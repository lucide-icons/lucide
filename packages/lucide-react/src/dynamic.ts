export {
  default as DynamicIcon,
  iconNames,
  type DynamicIconModule,
  type IconName,
} from './DynamicIcon';
export { default as dynamicIconImports } from './dynamicIconImports';
