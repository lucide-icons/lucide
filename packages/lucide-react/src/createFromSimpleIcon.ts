import createLucideIcon from './createLucideIcon';
import type { IconNode } from './types';
import type { SimpleIcon } from 'simple-icons';

const createIconNodeFromSimpleIcon = (simpleIcon: SimpleIcon): IconNode => [
  [
    'path',
    {
      d: simpleIcon.path,
      fill: 'currentColor',
    },
  ],
];

/**
 * Converts a simple-icon into a Lucide icon component
 * @param name The name of the icon (will be used as component name)
 * @param simpleIcon The simple-icon object
 * @returns A Lucide icon component
 */
const createFromSimpleIcon = (name: string, simpleIcon: SimpleIcon) =>
  createLucideIcon(name, createIconNodeFromSimpleIcon(simpleIcon));

export default createFromSimpleIcon;
