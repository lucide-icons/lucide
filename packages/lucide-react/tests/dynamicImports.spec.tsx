import { describe, it, expect } from 'vitest';
import { Suspense, lazy } from 'react';
import { render, waitFor } from '@testing-library/react';

import dynamicIconImports from '../src/dynamicIconImports';
import { LucideProps } from '../src/types';

describe('Using dynamicImports', () => {
  it('should render icons dynamically by using the dynamicIconImports module', async () => {
    interface IconProps extends Omit<LucideProps, 'ref'> {
      name: keyof typeof dynamicIconImports;
    }

    const Icon = ({ name, ...props }: IconProps) => {
      const LucideIcon = lazy(dynamicIconImports[name]);

      return (
        <Suspense fallback={null}>
          <LucideIcon {...props} />
        </Suspense>
      );
    };

    const { container, getByLabelText } = render(
      <Icon
        aria-label="smile"
        name="smile"
        size={48}
        stroke="red"
        absoluteStrokeWidth
      />,
    );

    await waitFor(() => getByLabelText('smile'));

    expect(container.innerHTML).toMatchSnapshot();
  });
});
