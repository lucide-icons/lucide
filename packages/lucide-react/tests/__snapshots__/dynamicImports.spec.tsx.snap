// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Using dynamicImports > should render icons dynamically by using the dynamicIconImports module 1`] = `
<svg xmlns="http://www.w3.org/2000/svg"
     width="48"
     height="48"
     viewbox="0 0 24 24"
     fill="none"
     stroke="red"
     stroke-width="1"
     stroke-linecap="round"
     stroke-linejoin="round"
     class="lucide lucide-smile"
     aria-label="smile"
>
  <circle cx="12"
          cy="12"
          r="10"
  >
  </circle>
  <path d="M8 14s1.5 2 4 2 4-2 4-2">
  </path>
  <line x1="9"
        x2="9.01"
        y1="9"
        y2="9"
  >
  </line>
  <line x1="15"
        x2="15.01"
        y1="9"
        y2="9"
  >
  </line>
</svg>
`;
